### DISCUSSION TITLE
Exclude functions and attributes from table of contents

### DISCUSSION THREAD

**Discussion by KevinMusgrave** [mkdocstrings/mkdocstrings#411] [Q&A] [answered]
The table of contents for one of my pages looks like this

![image](https://user-images.githubusercontent.com/2314891/160402914-b3b9d1f2-9752-43a3-8cd6-f1e4f0df368e.png)

All those ```__init__``` entries make it difficult to read. In general, I only want the top-level heading entries to appear in the table of contents. Is there a way to do that?

**Comment by pawamoy** [comment, marked as answer]
Hello! Multiple options:

- the new handler has the [`merge_init_into_class` rendering option](https://mkdocstrings.github.io/python/reference/mkdocstrings_handlers/python/renderer/#mkdocstrings_handlers.python.renderer.PythonRenderer.default_config). If your module is just composed of classes with `__init__` methods, this could help. There's also the `ignore_init_summary` docstring option that can help in that case ([Griffe's config](https://github.com/mkdocstrings/griffe/blob/master/mkdocs.yml#L83), [docs](https://mkdocstrings.github.io/griffe/reference/griffe/docstrings/google/#griffe.docstrings.google.parse)).
- with the legacy handler, you can exclude `__init__` methods from rendering, however I believe you *want* to render them but also to hide them from ToC. Well, that's not possible, and probably won't be made possible in the future.
- set a depth of 1 (or 2?) with the [toc extension](https://python-markdown.github.io/extensions/toc/) `depth` option

I'll move this as a Q/A discussion :slightly_smiling_face:

**Reply by KevinMusgrave** [reply]
```toc_depth: 2``` did the job, thanks!

**Reply by ssbarnea** [reply]
I am afraid that toc_depth does not help here because I have lots of methods that I want to document where I do have lots of others that I want to hide from toc, examples:

- hiding `__init__` because the main class name link is more than enough
- hiding most dunder methods such `__eq__`, `__hash__`, `__lt__`, `__deepcopy__`, `__getattribute__ `, `__getitem__ ` ......

The reality is that we need a list of method names to hide, so we can customize what gets rendered or not.

**Reply by pawamoy** [reply]
Do you want to hide them from the ToC, or to hide them completely?

**Reply by ssbarnea** [reply]
Good question! I think that the priority would be to hide them from TOC. Still, probably hiding them from documentation would not be so bad.

I wonder if there is a way to declare from within the docstrings, if you want it document it or not. 

For example I do see a use-case for documenting behavior related to the dunder method, but this kind of documentation might also be included at class-level docstrings. It does not make a lot of sense to document equality signature, but I could write a paragraph about the fact that the class object can be used in `==` expressions.

So, I am open to suggestions here. Probably I also be able to have a practical example to show.

**Reply by pawamoy** [reply]
To hide objects from the docs (and therefore from the ToC), you have the `members` and `filters` options at your disposal. I believe hiding them from the ToC only is possible as well: could you open a new feature request for that?

**Reply by pawamoy** [reply]
I recall someone using a single line of CSS to hide entries from the ToC. Something to keep in mind.
